x-pack/legacy/plugins/ingest/server/libs/adapters/datasource/default.ts

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { isRight, Right } from 'fp-ts/lib/Either';
+import { PathReporter } from 'io-ts/lib/PathReporter';
+import { ListOptions } from '../../../../../fleet/server/repositories/agents/types';
+import { FrameworkUser } from '../framework/adapter_types';
+import { SODatabaseAdapter } from '../so_database/default';
+import { RuntimeStoredDatasource, StoredDatasource } from './adapter_types';
+
+export class DatasourceAdapter {
+  constructor(private readonly so: SODatabaseAdapter) {}
+
+  public async create(
+    user: FrameworkUser,
+    datasource: StoredDatasource,
+    options?: { id?: string }
+  ): Promise<StoredDatasource> {
+    const newSo = await this.so.create<any>(
+      user,
+      'datasources',
+      (datasource as any) as StoredDatasource,
+      options
+    );
+
+    return {
+      id: newSo.id,
+      ...newSo.attributes,
+    };
+  }
+
+  public async get(user: FrameworkUser, id: string): Promise<StoredDatasource | null> {
+    const datasourceSO = await this.so.get<any>(user, 'datasources', id);
+    if (!datasourceSO) {
+      return null;
+    }
+
+    if (datasourceSO.error) {
+      throw new Error(datasourceSO.error.message);
+    }
+
+    const datasource = {
+      id: datasourceSO.id,
+      ...datasourceSO.attributes,
+    };
+
+    const decoded = RuntimeStoredDatasource.decode(datasource);
```

Comment:
Do we really need to validate data coming from saved object?

Reply:
this pattern was copied from policies api code, I'm not sure what the rationale was. let's check with Matt next week